video/out/vo_sixel.c

```diff
@@ -0,0 +1,429 @@
+    static unsigned int average_b = 0;
+    static int previous_histgram_colors = 0;
+    int histgram_colors = 0;
+    int palette_colors = 0;
+    unsigned char const* palette;
+
+    histgram_colors = sixel_dither_get_num_of_histogram_colors(priv->testdither);
+
+    if (priv->dither == NULL)
+        goto detected;
+
+    /* detect scene change if number of colors increses 20% */
+    if (previous_histgram_colors * 6 < histgram_colors * 5)
+        goto detected;
+
+    /* detect scene change if number of colors decreses 20% */
+    if (previous_histgram_colors * 4 > histgram_colors * 5)
+        goto detected;
+
+    palette_colors = sixel_dither_get_num_of_palette_colors(priv->testdither);
+    palette = sixel_dither_get_palette(priv->testdither);
+
+    /* compare color difference between current
+     * palette and previous one */
+    for (i = 0; i < palette_colors; i++) {
+        r += palette[i * 3 + 0];
+        g += palette[i * 3 + 1];
+        b += palette[i * 3 + 2];
+    }
+    score = (r - average_r) * (r - average_r)
+          + (g - average_g) * (g - average_g)
+          + (b - average_b) * (b - average_b);
+    if (score > priv->opts->threshold * palette_colors
+                             * palette_colors)
+        goto detected;
+
+    return 0;
+
+detected:
+    previous_histgram_colors = histgram_colors;
+    average_r = r;
+    average_g = g;
+    average_b = b;
+    return 1;
+}
+
+static SIXELSTATUS prepare_static_palette(struct vo* vo)
+{
+    struct priv* priv = vo->priv;
+
+    if (priv->dither)
+        sixel_dither_set_body_only(priv->dither, 1);
+    else {
+        priv->dither = sixel_dither_get(BUILTIN_XTERM256);
+        if (priv->dither == NULL)
+            return SIXEL_FALSE;
+        sixel_dither_set_diffusion_type(priv->dither, priv->opts->diffuse);
+    }
+    return SIXEL_OK;
+}
+
+static SIXELSTATUS prepare_dynamic_palette(struct vo *vo)
+{
+    SIXELSTATUS status = SIXEL_FALSE;
+    struct priv *priv = vo->priv;
+
+    /* create histgram and construct color palette
+     * with median cut algorithm. */
+    status = sixel_dither_initialize(priv->testdither, priv->buffer,
+                                     priv->opts->width, priv->opts->height, 3,
+                                     LARGE_NORM, REP_CENTER_BOX,
+                                     QUALITY_LOW);
+    if (SIXEL_FAILED(status))
+        return status;
+
+    if (detect_scene_change(vo)) {
+        if (priv->dither)
+            sixel_dither_unref(priv->dither);
+
+        priv->dither = priv->testdither;
+        status = sixel_dither_new(&priv->testdither, priv->opts->reqcolors, NULL);
+
+        if (SIXEL_FAILED(status))
+            return status;
+
+        sixel_dither_set_diffusion_type(priv->dither, priv->opts->diffuse);
+    } else
+        sixel_dither_set_body_only(priv->dither, 1);
+
+    return status;
+}
+
+static int resize(struct vo *vo)
+{
+    struct priv *priv = vo->priv;
+
+    talloc_free(priv->buffer);
```

Comment:
It seems that `priv->buffer`, `priv->dither`, and `priv->testdither` are always deallocated together, maybe move this sequence to its own function?

Also, is this called on terminal resize? or mpv video resize for whatever reason?

What happens on terminal resize? I don't think i've seen code which handles it.

Reply:
`priv->dither` is deallocated during the dynamic palette computation as well. Can club them in a function.

So terminal resize is not handled because mpv doesn't detect it (resize callback is not called). Even the behaviour is the same for tct and caca. We'd have to check each time in draw frame, or maybe periodically once per second or so to detect if terminal size has been changed and run resize accordingly, but not sure if including this check in every draw_image call would be optimal way of doing it. Probably resize is called if zoom or some other keybinding to change the input resolution is triggered (don't know how to trigger this).

Reply:
I think better way would be to add keybindings for zoom and shifts horizontal and vertical and probably a reset/optimum_screen keybindng which optimally computes using TIOCGWINSZ ioctl and which calls resize callback function. Let the user then accordingly zoom and move seems better option to me.

Reply:
OK, let's forget terminal resize for now. Maybe at a later stage we'll be able to handle it with sixel and tct.

Reply:
`priv->buffer` could possibly be deallocated automatically via talloc (check if the vo already has a talloc context).